http: add `http.createStaticServer` (#45096)
nodejs/node: doc/api/http.md

server.listen(8000);
```

> ## `http.createSimpleServer([directory[, port[, host]]])`

Reviewer: Bikeshed: s/Simple/Static or s/Simple/File

Reviewer: I'd prefer the argument here to be an `options` object that would allow us to more easily extend later, e,g,

```js
http.createStaticServer({
  root: '/the/directory',
  port: 1234,
  host: 'localhost'
})
```

<!--YAML
added: REPLACEME
-->

> Stability: 0 - Experimental

> * `directory` {string|URL} Root directory from which files would be serve.

Reviewer 2: ```suggestion
* `directory` {string|URL} Root directory from which files would be served.
```

  **Default:** `process.cwd()`.
* `port` {number}
* `host` {string} **Default:** `'localhost'`

* Returns: {http.Server}

Start a TCP server listening for connections on the given `port` and `host`, and

> serve statically local files, using `directory` as the root. Please note that

Reviewer 2: ```suggestion
serve static local files, using `directory` as the root. Please note that
```

> when specifying a `host` other than `localhost`, you are exposing you local file

Reviewer 2: ```suggestion
when specifying a `host` other than `localhost`, you are exposing your local file
```

system to all the machines that can connect to your computer.

If `port` is omitted or is 0, the operating system will assign an arbitrary

> unused port, which will be output the standard output.

Reviewer 2: I'm unsure what this is trying to say here, it's output will be the standard output or it'll output whatever comes from the standard output?

```suggestion
unused port, which it's output will be the standard output.
```

or whatever that fits correctly here.

Also accessible via `require('node:http/server')`.

```bash
node -r node:http/server # starts serving the cwd on a random port

# To start serving on the port 8080 using /path/to/dir as the root:

> node -r node:http/server /path/to/dir --port 8080

Reviewer: Not really a fan of building in a preload like this. Someone can easily use the `-pe` syntax to achieve this if they wanted, e.g. `node -pe "require('http').createStaticServer(...)`

Author: One reason to use `-r` is you can get autocomplete when selecting the root directory, and it's easier to offer shorter versions of flags (i.e. I'd rather type `node -r http/server . -p 8080 -h 0.0.0.0` than `node -e 'http.createStaticServer({port: 8080, host: '0.0.0.0'})'`). Currently this PR offers both syntax, so I guess I should also document that, and folks can chose the syntax they prefer, wdyt?

> # Same, but exposing your local file system to the whole IPv4 network:

Reviewer 2: ```suggestion
# Same as above, but exposing your local file system to the whole
# IPv4 network:
```

server.listen(8000);
```

## `http.createSimpleServer([directory[, port[, host]]])`

<!--YAML
added: REPLACEME
-->

> Stability: 0 - Experimental

* `directory` {string|URL} Root directory from which files would be serve.
* `port` {number}
* `host` {string}

Start a TCP server listening for connections on the given `port` and `host`, and
serve statically local files, using `directory` as the root. Please note that
when specifying a `host` other than `localhost`, you are exposing you local file
system to all the machines that can connect to your computer.

If `host` is omitted, the server will accept connections on the
[unspecified IPv6 address][] (`::`) when IPv6 is available, or the

> [unspecified IPv4 address][] (`0.0.0.0`) otherwise.

Reviewer 3: Why not use 127.0.0.1 and the equivalent IPv6 by default?

Author: I copied this from https://github.com/nodejs/node/blob/e23c256594c291dd2da481512e2dc1d0e4421375/doc/api/net.md#L544-L554

I've considered using `'localhost'` as a default value, but thought maybe consistency would be preferable, happy to change that back if others prefer it.

Reviewer 3: I'd prefer the default to be safe in this case (avoid potentially exposing a local folder to the entire internet).

server.listen(8000);
```

## `http.createStaticServer([options])`

<!--YAML
added: REPLACEME
-->

> Stability: 0 - Experimental

* `options` {Object}
  * `directory` {string|URL} Root directory from which files would be served.
    **Default:** `process.cwd()`.
  * `port` {number}
  * `host` {string} **Default:** `'localhost'`
  * `mimeOverrides` {Object} Dictionary linking file extension to a MIME string,
    to override or extend the built-in ones.

>   * `serveDotFiles` {boolean} **Default:** `false`

Reviewer: Bikeshed: `s/serveDotFiles/allowDotFiles`

Just an alternative (non-blocking) suggestion, as an option to consider... define this as a custom `filter()` function where the default implementation filters out dot files... essentially something like:

```js
http.createStaticServer({
  directory: './files',
  port: 1234,
  filter(pathName) {
    // return true if the pathName is acceptable.
    // if filter is not provided, the default would block dot files or paths with dot segments (e.g. both `/.foo`, `/.foo/bar` would be rejected) 
  };
});
```

server.listen(8000);
```

## `http.createStaticServer([options])`

<!-- YAML
added: REPLACEME
-->

> Stability: 1 - Experimental. The implementation lacks support for partial
> responses (Ranges) and conditional-GET negotiation (If-Match,
> If-Unmodified-Since, If-None-Match, If-Modified-Since).

* `options` {Object}

>   * `directory` {string|URL} Root directory from which files would be served.

Reviewer 3: ```suggestion
  * `directory` {string|URL} Root directory from which files will be served.
```

Or "should" / "must"? /cc @nodejs/documentation

Reviewer 3: ^

    **Default:** `process.cwd()`.
  * `port` {number}
  * `host` {string} **Default:** `'localhost'`
  * `mimeOverrides` {Object} Dictionary linking file extension to a MIME string,
    to override or extend the built-in ones.
  * `filter` {Function|null} should be a function that accepts two arguments and
    returns a value that is coercible to a {boolean} value. When `null`, no
    files are filtered. **Default:** filters all dot files.
  * `log` {Function|null} called when sending a response to the client.

>     **Default:** `console.log`.

Reviewer 3: We have to document what the function will be called with.

Author: ```suggestion
  * `log` {Function|null} called when sending a response to the client.
    **Default:** `console.log`.
      * `statusCode` {integer} The status code that is sent to the client.
      * `url` {string} the (origin-relative) URL that was requested.
```

  * `onStart` {Function} called the server starts listening to requests.
    **Default:** logs the URL to the console.

* Returns: {http.Server}

Start a TCP server listening for connections on the given `port` and `host`, and
serve static local files, using `directory` as the root.
When specifying a `host` other than `localhost`, you are exposing your local
file system to all the machines that can connect to your computer.

If specified and not `null`, `filter` will be called with two arguments: the

> first one if the request URL string (the URL that is present in the actual HTTP

Reviewer 3: ```suggestion
first one is the request URL string (the URL that is present in the actual HTTP
```

request), and the second one is the `file:` `URL` that was generated from the
base directory and the request URL. If the function returns a falsy value, the

> request will be blocked.

Reviewer 3: I think we should document how it will be blocked

Author: ```suggestion
base directory and the request URL. If the function returns a falsy value, the
server will respond with a 403 HTTP error.
```
